More efficient management of code fragments

xavierleroy opened the pull request:
The OCaml runtime system keeps track of "code fragments", that is, areas of memory containing byte- or native- compiled code.  Typically, there is one fragment for the main, statically-linked program, and one fragment for each dynamically-loaded module.

Cdoe fragments play a role in marshaling and unmarshaling of function closures, but also in the interface with the bytecode debugger.  Future uses are being considered for hashing (https://github.com/ocaml/ocaml/pull/9648#issuecomment-640590514) and for scanning the bytecode interpreter stack.

Currently, the management of code fragments is spread over several runtime modules, and uses inefficient data structures (extensible arrays) with search time linear in the number of code fragments.

This PR groups together all operations on code fragments in a new runtime module `codefrag.c` with a well-defined interface in `<caml/codefrag.h>`.  It uses the new `skiplist.c` module to implement "search code fragment by PC" in logarithmic time.

As usual, this PR is best reviewed commit by commit.

xavierleroy committed:
codefrag: a new runtime module to work with code fragments

This module collects all the operations on code fragments performed
in various places of the runtime systems.  Applies both to bytecode
and to native code.

It also improves the handling of digests: now it is possible to
mark a code fragment as "no digest" i.e. not marshal-able.

xavierleroy committed:
Use the new "codefrag" runtime module

Replace the previous handling of code fragments with calls to the
functions provided by the "codefrag" runtime module.

Benefits include faster lookups of code fragments by PC
(logarithmic instead of linear in the number of code fragments).

xavierleroy commented:
Bringing this to @jhjourdan 's attention.

xavierleroy commented on runtime/meta.c:
```diff
@@ -93,25 +94,26 @@ CAMLprim value caml_reify_bytecode(value ls_prog,
   } else {
     /* | None -> */
-    cf->digest_computed = 0;
+    digest_kind = DIGEST_LATER;
```
This None case is used for bytecode produced for phrases entered at the REPL (as opposed to bytecode loaded from .cmo files).  `DIGEST_LATER` matches the previous behavior: it makes it possible to marshal functions entered at the REPL, and unmarshal them in the same `ocaml` session, but unmarshaling in a different `ocaml` session usually doesn't work.  `DIGEST_IGNORE` would prevent these functions from being marshaled.

stedolan commented on runtime/codefrag.c:
```diff
@@ -0,0 +1,126 @@
+  return pos;
+}
+
+void caml_remove_code_fragment(struct code_fragment * cf)
```
When a code fragment is removed, the fragnums of all subsequent code fragments change. Won't this confuse the debugger, which still has the old fragnums in its `ev_frag`s?

xavierleroy replied:
I was careful to reproduce exactly the behavior of the current code:
https://github.com/ocaml/ocaml/blob/2e4d1ecc12cfc6daa3b87e0400aed12544d877a9/runtime/meta.c#L156-L159

The effect of `caml_ext_table_remove` is to decrease the fragment numbers of all fragments "above" the one being removed.

I was hoping there was a corresponding adjustment to fragment numbers on the ocamldebug side, but I cannot find it anywhere in the ocamldebug sources.

@jhjourdan @whitequark : can you please clarify what should happen to fragment numbers when a fragment is unloaded?

whitequark replied:
I unfortunately no longer recall OCaml-related specifics as I worked on this too long ago, but that being said, since the debugger is reliably notified of the lifetime of fragment numbers, IMO they can be safely reused.

xavierleroy replied:
Thanks for your quick reply, however you may be missing the point.

Say we have 3 code fragments in use, with fragment numbers 0 1 2.

Say we unload the fragment number 1.

`caml_ext_table_remove` "slides" fragment number 2 at slot number 1.  So, from now on, the code that use to have fragment number 2 has fragment number 1.

The debugger is notified that fragment number 1 was unloaded.  However, I don't see where the debugger replaces fragment number 2 by fragment number 1.

whitequark replied:
Oh, indeed I am missing the point. I suspect it's just broken; I remember that back when I worked on it, code loading was essentially nonfunctional in both ocamldebug and the emacs frontend, so it wouldn't surprise me if no one ever verified this code path.

stedolan replied:
The issue isn't just ID reuse, it's that the IDs of code fragments change midway through their lifetimes as they get shifted around by other code fragments being removed.

Maybe we should use a skiplist rather than an ext_table to map fragment numbers to code_fragments? That way, we can give code fragments unique immutable IDs.

gasche replied:
If I understand correctly, code fragment numbers where important in the old world where we could only (easily) address PCs by knowing their code fragment. If we can now lookup PCs efficiently, do we still need the code fragment number in the debugger? (I suppose that there is no risk that PCs would overlap?)

jhjourdan replied:
Recall that in the current implementation, code fragments can only be unloaded if they correspond to code written directly in the toplevel (and ony if this code does not include functions).

In this situation, it is very likely that the unloaded code fragment is the last one (well, except if this piece of code dynamically loads a plugin). 

So, the bug @xavierleroy is mentionning can only occur is very rare situations : when loading a plugin when running a piece of code from the toplevel, when this toplevel is watched by ocamldebug.

Now, if we want to avoid this, we can still assign code fragment unique identifiers by incrementing a counter each time a code fragment is loaded.

xavierleroy replied:
All right.  I rewrote the code by 1- not changing fragment numbers, 2- reclaiming space in caml_code_fragments_table only when the removed fragment is the one most recently registered, and 3- leaving a hole in caml_code_fragments_table otherwise.

We could do better by maintaining a free list of holes in caml_code_fragments_table, to be reused at the next registration, or by switching to a skip list indexed by fragment numbers.  But this should do for now.

xavierleroy replied:
> If we can now lookup PCs efficiently, do we still need the code fragment number in the debugger? (I suppose that there is no risk that PCs would overlap?)

We could use the start PC as code fragment identifier, I guess.  We would need to use nativeint instead of int in the debugger and in the protocol, though.

xavierleroy replied:
More generally: I was trying to not change the debugger and the communication protocol, just improve performance of the "find fragment by PC" operation.

jhjourdan commented on runtime/codefrag.c:
```diff
@@ -0,0 +1,126 @@
+  if (caml_skiplist_find_below(&caml_code_fragments_list,
+                               (uintnat) pc, &key, &data)) {
+    cf = (struct code_fragment *) data;
+    if (cf->code_start <= pc && pc < cf->code_end) return cf;
```
Isn't the first condition of the test already true?

damiendoligez replied:
It could be replaced by an assertion to make it clear that `pc` is in the right range.

xavierleroy replied:
Took me a while to get your point.  See commit dc6c33fd7 .

xavierleroy committed:
caml_remove_code_fragment: preserve all fragment numbers

The previous code, like the OCaml 4.11 implementation,
was sliding fragments over the removed fragment,
causing some fragment numbers to decrease by 1.

It turns out the debugger is not set up to handle these changes
in fragment numbers.

Avoid the problem by reclaiming space in caml_code_fragments_list
only in the common case where the fragment being removed
is most recently inserted fragment.

jhjourdan commented on runtime/codefrag.c:
```diff
@@ -0,0 +1,126 @@
+{
+  int i;
+
+  for (i = caml_code_fragments_table.size - 1; i >= 0; i--) {
+    struct code_fragment * cf = caml_code_fragments_table.contents[i];
+    unsigned char * d = caml_digest_of_code_fragment(cf);
+    if (d != NULL && memcmp(digest, d, 16) == 0) return cf;
+  }
```
This loop is going to be executed each time a closure is deserialized. Hence, this might be performance sensitive. Why not also using a skiplist indexed on e.g., the first word of the MD5 sum? Since the skiplist implementation is already there, that should not be so difficult.

jhjourdan replied:
Now that I think about it more, I understand that this might be more difficult than expected since several code fragments may share the same first word of digest.... We would need to patch the skiplist implementation to allow multiple bingdings for a key.

xavierleroy replied:
You can link the code fragments that have a collision on the first word of the digest, like in a dynamic hash table.  But there's an additional wrinkle for code fragments whose digest is computed lazily.

jhjourdan replied:
Why is this a problem? These code fragments would simply not appear in the skiplist.

xavierleroy replied:
> These code fragments would simply not appear in the skiplist.

The DIGEST_IGNORE fragments would never be put in the skiplist (indexed by digests), yes.  The DIGEST_LATER fragments would need to be digested and inserted in the skiplist at the first call to `caml_find_code_fragment_by_digest`.

xavierleroy replied:
The things you make me do, squared...  Have a look at commit 602aa0489 !

xavierleroy replied:
Don't have a look at commit 602aa04 : removal is wrong.

xavierleroy replied:
In the end I backtracked.  Too much code needed to implement the "by digest" skiplist, esp. w.r.t. removal.

jhjourdan reviewed (approved):
This looks good to me.

Of course, there is the problem of code fragment numbers for the debugger.

I have another concern about the performance of the search by digest (see comments).

jhjourdan commented on runtime/codefrag.c:
```diff
@@ -64,18 +64,18 @@ int caml_register_code_fragment(char * start, char * end,
+    /* Just mark the table entry invalid.  Later, we could maintain
+       a free list of empty entries in caml_code_fragments_table
+       so as to reuse them on the next caml_register_code_fragment.
+       Or replace the table by a skiplist. */
```
Is there any major complication in using a skiplist? If the code is no more complicated, then let's use the alternative which works better from the beginning (i.e., use a skiplist).

xavierleroy replied:
The things you make me do...  See commit 0107f395f .

jhjourdan replied:
I'm honored to make you do these things ;)

jhjourdan replied:
I had a look at the code : this seems good to me.

xavierleroy committed:
codefrag.c: use a skiplist indexed by fragment number

By popular demand.

damiendoligez commented on runtime/codefrag.c:
```diff
@@ -0,0 +1,157 @@
+  caml_skiplist_insert(&code_fragments_by_pc,
+                       (uintnat) start, (uintnat) cf);
+  caml_skiplist_insert(&code_fragments_by_num,
+                       cf->fragnum, (uintnat) cf);
```
```suggestion
                       (uintnat) cf->fragnum, (uintnat) cf);
```
I think it's better to have an explicit rather than implicit cast.

xavierleroy replied:
All right.  See commit dc6c33fd7

damiendoligez commented on runtime/codefrag.c:
```diff
@@ -0,0 +1,157 @@
+{
+  caml_skiplist_remove(&code_fragments_by_pc, (uintnat) cf->code_start);
+  caml_skiplist_remove(&code_fragments_by_num, cf->fragnum);
+  caml_stat_free(cf);
```
That can't be right. Surely you also have to remove it from `code_fragments_by_digest` and (or?) `code_fragments_recently_added`. If I understand correctly, the fragment is very likely to be at (or near) the head of `code_fragments_recently_added`, so it shouldn't be too costly to just walk down the list.

xavierleroy replied:
Yes, I forgot to implement removal in the "code fragment by digest" implementation.  I did, and it's a lot of nasty code.  You can see it here: https://github.com/xavierleroy/ocaml/blob/code-fragments-super-duper/runtime/codefrag.c .  But it is too much code, really, so I reverted the last commit: the `code_fragments_by_digest` skip list is not there, and `caml_find_fragment_by_digest` performs a linear search, like in the good old times.

damiendoligez commented on runtime/codefrag.c:
```diff
@@ -0,0 +1,157 @@
+
+unsigned char * caml_digest_of_code_fragment(struct code_fragment * cf)
+{
+  if (cf->digest_status == DIGEST_IGNORE)
```
This is a pattern-matching on `digest_status` so a `switch` statement would be clearer (and possibly faster).

xavierleroy replied:
I wrote it with a switch initially but it wasn't that nice.

damiendoligez commented on runtime/codefrag.c:
```diff
@@ -0,0 +1,157 @@
+  if (! caml_skiplist_find(&code_fragments_by_digest, key, &data))
+    return NULL;
+  for (cf = (struct code_fragment *) data; cf != NULL; cf = cf->next) {
+    CAMLassert(cf->digest_status == DIGEST_NOW ||
+               cf->digest_status == DIGEST_PROVIDED);
```
```suggestion
    CAMLassert(cf->digest_status == DIGEST_PROVIDED);
```
I don't think it can ever be `DIGEST_NOW`.

damiendoligez commented on runtime/codefrag.c:
```diff
@@ -0,0 +1,157 @@
+    /* Stop early if we find the digest we're looking for. */
+    if (memcmp(digest, d, 16) == 0) return cf;
+  }
+  /* Look the digest up in code_fragments_by_digest */
```
You could do this first, and start working on `code_fragments_recently_added` only if it fails.

damiendoligez commented on runtime/extern.c:
```diff
@@ -713,15 +713,15 @@ static void extern_rec(value v)
-    }
+    digest = (const char *) caml_digest_of_code_fragment(cf);
+    if (digest == NULL)
+      extern_invalid_argument("output_value: private function");
```
This is a new behaviour, right? Would you care to document it?

xavierleroy replied:
It is a new behavior, although it cannot be exercised yet.  (No DIGEST_IGNORE fragment registrations yet.) Depending on how DIGEST_IGNORE will be used later, this could almost be an assertion `digest != NULL`.  At this point I'll stick with an Invalid_argument exception but I don't feel like documenting it.

damiendoligez commented on runtime/startup_nat.c:
```diff
@@ -79,12 +78,9 @@ static void init_static(void)
-  caml_ext_table_add(&caml_code_fragments_table, cf);
+  caml_register_code_fragment(caml_code_area_start,
+                              caml_code_area_end,
+                              DIGEST_LATER, NULL);
```
I'm curious: why do you use `LATER` here and `NOW` in byte-code?

xavierleroy replied:
Long story!

In bytecode we want the digest of a piece of code to be the same across different architectures, so that these architectures can exchange function pointers.  However, before execution the bytecode is changed: endianness conversion, threading.  So we want to compute the digest before those changes, i.e. NOW.  It's not expensive, since the whole bytecode is being touched to do the endianness conversion and the threading at any rate.

In native code, the code does not change between program start-up and the first marshaling/unmarshaling operation that needs to know its digest.  So we can compute the digest on demand.  Most ocamlopt-compiled programs will never compute the digest!  Computing it at program start-up would be costly, as we would be touching all the pages of the code at once.

damiendoligez reviewed (changes requested):
Seems good in general, but there's one thing that looks wrong to me.

xavierleroy committed:
Address review comments

xavierleroy committed:
Changes entry for #9654

xavierleroy commented:
Thanks @jhjourdan and @damiendoligez for the reviews.  I think I addressed all the comments.  I got cold feet about the skip list indexed by the top bits of the digest: it is a lot of code, and we haven't heard complaints yet about `input_value` being slow because of linear search of code digests.  If we ever, I promise to revisit this idea.  Right now, what matters to me is that "find fragment by PC" is always reasonably efficient even if we have an unreasonable number of code fragments.

jhjourdan commented:
Indeed, the code for removing from the digests skiplist is rather involved. Let's postpone this.

xavierleroy commented:
I consider that the changes required by @damiendoligez have been performed and will now merge this PR in its present state.

gasche commented:
I have been in touch with @hackwaly who is encountering performance issues in [ocamlearlybird](https://github.com/hackwaly/ocamlearlybird) when debugging bytecode executables on recent OCaml versions. I took the initative to backport this PR into 4.11 (=> 751f965f801cfc9ba0c3be12d0fa9edd200d8073), which in turn required backporting #9635 (=> fb79bd82499f8a90efb2e2bdb1bb57b62f4f5262) and #9660 (=> 286e21cf4ac536857b43a580d79a36d18f2241a6).

(The backport was immediate.)
